File: php/tests/ReconcileTest.php
<?php

declare(strict_types=1);

namespace Pagou\Examples\Tests;

use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Psr7\Response;
use Pagou\Examples\Config;
use Pagou\Examples\HttpClient;
use Pagou\Examples\Reconcile;
use PHPUnit\Framework\TestCase;

final class ReconcileTest extends TestCase
{
    /** @param list<Response> $queue */
    private function client(array $queue): HttpClient
    {
        $config = new Config('sandbox', 'https://api.sandbox.pagou.ai', 'test', null, null, 1000, 0);
        $guzzle = new GuzzleClient(['handler' => HandlerStack::create(new MockHandler($queue)), 'http_errors' => false]);
        return new HttpClient($config, $guzzle);
    }

    private static function json(int $status, array $body): Response
    {
        return new Response($status, ['Content-Type' => 'application/json'], json_encode($body));
    }

    public function testFulfillsOnSettledStatuses(): void
    {
        self::assertSame(Reconcile::FULFILL, Reconcile::decideFulfillment('paid'));
        self::assertSame(Reconcile::FULFILL, Reconcile::decideFulfillment('captured'));
    }

    public function testWaitsOnNonTerminalStatuses(): void
    {
        self::assertSame(Reconcile::WAIT, Reconcile::decideFulfillment('pending'));
        self::assertSame(Reconcile::WAIT, Reconcile::decideFulfillment('three_ds_required'));
        self::assertSame(Reconcile::WAIT, Reconcile::decideFulfillment('processing'));
    }

    public function testCancelsOnTerminalFailures(): void
    {
        self::assertSame(Reconcile::CANCEL, Reconcile::decideFulfillment('expired'));
        self::assertSame(Reconcile::CANCEL, Reconcile::decideFulfillment('refused'));
        self::assertSame(Reconcile::CANCEL, Reconcile::decideFulfillment('canceled'));
    }

    public function testFetchesTransactionAndReturnsDecision(): void
    {
        $client = $this->client([self::json(200, ['success' => true, 'requestId' => 'r', 'data' => ['id' => 'tx_1', 'status' => 'paid']])]);
        $result = Reconcile::transaction('tx_1', $client);
        self::assertSame(Reconcile::FULFILL, $result['decision']);
        self::assertSame('paid', $result['transaction']['status']);
    }

    public function testReturnsNullWhenTransactionDoesNotExist(): void
    {
        $client = $this->client([self::json(404, ['message' => 'not found'])]);
        self::assertNull(Reconcile::transaction('missing', $client));
    }
}

File: ruby/tests/reconcile_test.rb
# frozen_string_literal: true

require_relative "test_helper"

class DecideFulfillmentTest < Minitest::Test
  def test_fulfills_only_on_settled_statuses
    assert_equal :fulfill, Pagou.decide_fulfillment("paid")
    assert_equal :fulfill, Pagou.decide_fulfillment("captured")
  end

  def test_waits_on_non_terminal_statuses
    assert_equal :wait, Pagou.decide_fulfillment("pending")
    assert_equal :wait, Pagou.decide_fulfillment("three_ds_required")
    assert_equal :wait, Pagou.decide_fulfillment("processing")
  end

  def test_cancels_on_terminal_failures
    assert_equal :cancel, Pagou.decide_fulfillment("expired")
    assert_equal :cancel, Pagou.decide_fulfillment("refused")
    assert_equal :cancel, Pagou.decide_fulfillment("canceled")
  end
end

class ReconcileTransactionTest < Minitest::Test
  def test_fetches_the_transaction_and_returns_a_decision
    client = StubClient.new(FakeResponse.new(200,
                                             { "success" => true, "requestId" => "r",
                                               "data" => { "id" => "tx_1", "status" => "paid" }, }))
    result = Pagou.reconcile_transaction("tx_1", client)
    assert_equal :fulfill, result[:decision]
    assert_equal "paid", result[:transaction]["status"]
  end

  def test_returns_nil_when_the_transaction_does_not_exist
    client = StubClient.new(FakeResponse.new(404, { "message" => "not found" }))
    assert_nil Pagou.reconcile_transaction("missing", client)
  end
end
